Weighted gap mortar mechanical contact implementation

lindsayad opened the pull request:
Closes #16961 

@recuero

lindsayad committed:
Execute mortar constraints before residual/matrix assembly

lindsayad committed:
Two-pass implementation of weighted gap LM mortar constraint

Closes #16961

lindsayad committed:
Demonstrate how to use two-pass objects

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/695016/) on 9a831c3 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/17189/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/17189/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format aed72321f3e543aa40a2e293f5d597360e89d832`

lindsayad committed:
Prevent repeat failed projections of primary nodes

lindsayad committed:
If reached max iterations, consider projection failed

lindsayad committed:
Don't call dof_number if not n_dofs

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/695193/) on 652b298 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/17189/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/17189/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format aed72321f3e543aa40a2e293f5d597360e89d832`

lindsayad committed:
Mixed basis test

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/696374/) on ff7ea7e wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/17189/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/17189/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format aed72321f3e543aa40a2e293f5d597360e89d832`

lindsayad committed:
Documentation

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/696549/) on a00b130 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/17189/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/17189/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format aed72321f3e543aa40a2e293f5d597360e89d832`

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/696677/) on 3d72c8e wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/17189/site)

This comment will be updated on new commits.

recuero commented on modules/contact/src/constraints/ApplyPenetrationConstraintLMMechanicalContact.C:
```diff
@@ -0,0 +1,210 @@
+  _connected_dof_indices.clear();
+  _connected_dof_indices.push_back(_var.nodalDofIndex());
+
+  _qp = 0;
```
I am failing to see how the Gauss point variable is changing (maybe it shouldn't) in the calls to the *Qp* methods to fill residuals and Jacobians.

lindsayad replied:
Yea `_qp` is kind of a misnomer here because we are not applying any quadrature rules. We are just evaluating secondary values at one location (the secondary node). (And for `NodeFaceConstraints` which work the the primary face, obviously we only work with one location over there as well...the location where the secondary node projects). But to have consistency with the moose data structures we need to index into our `MooseArrays` holding variable value data; in this case we know the data should have size 1, so `_qp` must be 0. Perhaps it would be more transparent if we didn't do anything with `_qp` and explicitly just indexed with 0?

recuero replied:
Thanks for the details! I get it now. I think it's fine how it is, just lack of familiarity on my side.

recuero commented on modules/contact/src/constraints/ApplyPenetrationConstraintLMMechanicalContact.C:
```diff
@@ -0,0 +1,210 @@
+      _projection_successful = true;
+      // We've stored the integrated gap in the residual vector
+      Real & integrated_gap = _node_to_integrated_gap[_current_node];
+      integrated_gap = _residual_copy(_dof_number) / _var.scalingFactor();
```
Here, the scaling factor applied to the integrated gap corresponds to the primary variables? I.e. `disp_x` or `disp_y`?

lindsayad replied:
No it's the scaling factor for the LM variable. When we compute the weighted gap in our mortar constraint and save it into the residual vector, the scaling factor for the LM variable gets applied via multiplication. To undo that math, e.g. to make sure that our integrated gap here is truly representative of the gap value and not the gap value times some arbitrary factor, we divide by the scaling factor.

recuero commented on modules/contact/src/constraints/ComputeWeightedGapLMMechanicalContact.C:
```diff
@@ -0,0 +1,89 @@
+        _primary_disp_y,
+        "It doesn't make any sense that we have a secondary displacement variable and not a "
+        "primary displacement variable");
+    _computing_gap_dependence = true;
```
`_computing_gap_dependence` just refers to computing its Jacobian, doesn't it? When wouldn't we want to use it?

lindsayad replied:
We probably do always want to compute the (approximate) Jacobian. In fact my tests never converged if I didn't have those approximate Jacobians added in.

I copied this logic over from the other class...but yea I think what we want to do is make the `secondary_disp_y` parameter required and then just always compute the gap dependence. :+1:

lindsayad commented on modules/contact/src/constraints/ApplyPenetrationConstraintLMMechanicalContact.C:
```diff
@@ -0,0 +1,210 @@
+bool
+ApplyPenetrationConstraintLMMechanicalContact::shouldApply()
+{
+  const bool has_dofs = _current_node->n_dofs(_sys.number(), _var.number());
```
If the LM doesn't have any dofs at the node, e.g. a first order LM on a mid-edge node, then we do not apply the constraint because there is no dof to constrain

lindsayad commented on modules/contact/src/constraints/ApplyPenetrationConstraintLMMechanicalContact.C:
```diff
@@ -0,0 +1,210 @@
+          _qp < _u_secondary.size(),
+          "qp is greater than the size of u_secondary in "
+          "ApplyPenetrationConstraintLMMechanicalContact. Check and "
+          "make sure that your Lagrange multiplier variable has the same order as the mesh");
```
Yea this assertion can go. Nice catch!

lindsayad committed:
Remove need to call NonlinearSystemBase::mortarConstraints twice

It's more transparent just to have a single call to `mortarConstraints`
that executes both undisplaced and displaced mortar constraints. The old
way was motivated by having computation of mortar constraints occur in
`constraintResiduals/constraintJacobians` but we now compute mortar
constraints ahead of those methods in order to ensure that mortar
contributions are assembled into the global residual/Jacobian before we
evaluate other constraints

lindsayad commented:
@dschwen @bwspenc, could I get a CCB review/approval on this? @recuero approved (before I autosquashed and dismissed his review)

recuero commented:
@lindsayad Fantastic results! Your weighted gap (`wgap`) mortar implementation generates the best expected outcome!

The image below is for a non-flat geometry, which happens to be circular. I can see two differentiating aspects over node on face and the nodal gap mortar enforcement:

- In an area where there is a geometric defect, the weighted gap reduces the maximum pressure, which makes sense because the finer we discretize the more nodes will share the contacting force. I think this is a feature of mortar that guarantees the passage of the contact patch test. There is a spatial shift, but I think this is probably fine and due to segment-matching effects.

- In a "regular" area where node on face and the old nodal LM mortar formulations generate an artificial and ugly kink in the  results, the weighted gap generates a flawless evolution of the contact force.

The distribution of the contact pressure around the defect may be improved if we use the mixed approach, I'll keep trying combinations.

Thanks for working on this!


![image](https://user-images.githubusercontent.com/14300313/110181382-e68f1500-7dc8-11eb-82e1-0644a65714f0.png)

lindsayad commented:
Very cool! What is the difference between "node-segment" and "node-segment kinematic"? It looks like node-segment (kinematic) is less oscillatory at the geometric defect than the mortar implementations. I would indeed be curious whether the mixed approach helps with that. It's known that saddle point problems (which you would have in a same-order LM approach when you are in contact) can be oscillatory/produce checkerboards. @jwpeterson showed a lot of this same-order instability in his technical report when looking at solution continuity, which is also saddle-point.

recuero commented:
"node-segment" is obtained using a penalty approach, as opposed to the kinematic approach. It doesn't add much info to the plot since the results are virtually identical.

recuero commented:
Using first-order LMs on a second-order mesh (`mixed`) does show improvement with the oscillatory behavior around the defect. But it makes convergence a lot more challenging, to the point that the simulation can't finish. May be related to the approximate AD Jacobians for mechanical contact? This behavior was already present in the previous mortar approaches.

![image](https://user-images.githubusercontent.com/14300313/110215945-9ae07800-7e69-11eb-8012-5ffb188108db.png)

lindsayad commented:
I could believe it's a Jacobian problem. Is it taking a lot of linear iterations? Would you share the inputs you are using for this (either here or privately) or paste the output of a failed solve with `-ksp_monitor_true_residual`?

lindsayad committed:
Solve nodal projection tighter

It doesn't make sense to me to solve the (non)linear projection system
with the same tolerance as the comparison tolerance we're going to use
when deciding, for example, whether we should have already established a
map between a secondary and primary node. In general we always do our
solve tolerances tighter than our comparison tolerances

lindsayad committed:
Restore single way of computing normal field for mortar

We revert to the initial John Peterson implementation of normal fields
where we use the secondary nodal normals both for projecting secondary
and primary nodes. This seems to avoid issues of different projection
behavior between the sides (duh)

lindsayad commented:
@recuero please try playing around with these most recent commits. Your input that you sent me converged within 3 nonlinear iterations for every timestep with SMP.

lindsayad commented:
You'll probably find some new issues, but I'm pretty stoked with how that input solved! I always like to think that we're about to usher in a new era of mortar...

lindsayad commented:
After 484dbef the mixed basis mortar test now has the new record for fewest nonlinear iterations to solve the bouncing block problem at 93.

lindsayad commented:
Mmm I guess that it was actually 93 before that commit too

recuero commented:
For a relatively fine problem, with the two commits you added yesterday (including the tighter tolerance), I see that:

- For a first order LM on a first order mesh, the number of nonlinear iterations go from 142 to 137. The number of linear iterations go from 1113 to 1028. All this points to improvement on first order meshes.
- For a first order LM on a second order mesh, I couldn't solve the problem without those two commits. With those two commits, the number of nonlinear iterations is 121, and the number of linear iterations is 1450, which I think it's pretty good for this problem.

recuero commented:
Now, even for a second-order mesh, the thermomechanical problem can be solved to tight tolerances. Big improvement!

lindsayad commented:
Great news!

tophmatthews commented:
Are you using thermal/mechanical mortar @recuero ?

recuero commented:
Yes. For this particular problem, the thermal part wasn't very problematic.

lindsayad commented:
> For a relatively fine problem, with the two commits you added yesterday (including the tighter tolerance), I see that:
> 
>     * For a first order LM on a first order mesh, the number of nonlinear iterations go from 142 to 137. The number of linear iterations go from 1113 to 1028. All this points to improvement on first order meshes.
> 
>     * For a first order LM on a second order mesh, I couldn't solve the problem without those two commits. With those two commits, the number of nonlinear iterations is 121, and the number of linear iterations is 1450, which I think it's pretty good for this problem.

Pretty interesting, but not totally unexpected to me, that the mixed-basis (second order disp, first order LM) shows less nl its (121) than the equal-order first-order scheme (137). This is consistent with my experience with other saddle-point problems in which the LM has to be of a different order than the primal var to have stability (unless you implement another method of stabilization, e.g. adding additional terms like is done with PSPG for incompressible Navier-Stokes and like we implement for mortar solution continuity) in the solution...and generally what I see is that stability in the solution (e.g. lack of or less oscillations) equates to stability in the (non)linear solve.

lindsayad committed:
Regold failing contact tests after changing primary node projection

We changed the projection method for primary nodes in a previous commit.
More specifically we now use the same normal field for projecting
primary nodes as we do for projecting secondary nodes, with the normal
field corresponding to interpolation of secondary nodal normals. This
change was made with the target of creating a consistent mortar mesh,
e.g as an example we should never run into the case where a primary node
projects onto a secondary node, but the secondary node doesn't project
onto the primary node (or visa versa). We ran into this issue quite a
bit in a fairly simple input that Antonio sent me, but the change to a
single normal projection field fixed the issue (or made it a lot better
at a minimum).

With the projection method changed, we have to regold some contact tests

lindsayad commented:
@dschwen @bwspenc I'm pretty happy with where we've gotten to, so feel free to review when able

fdkong commented on modules/contact/src/constraints/ComputeWeightedGapLMMechanicalContact.C:
```diff
@@ -0,0 +1,65 @@
+      {
+        DualRealVectorValue gap_vec = _phys_points_primary[_qp] - _phys_points_secondary[_qp];
+        // Here we're assuming that the user provided the x-component as the secondary/primary
+        // variable!
```
If we do such an assumption, do we have a error check for users' input?

lindsayad replied:
Good idea. I will add error checks

fdkong commented on modules/contact/test/tests/bouncing-block-contact/frictionless-weighted-gap-mixed-basis.i:
```diff
@@ -0,0 +1,173 @@
+    primary_subdomain = 4
+    secondary_subdomain = 3
+    variable = normal_lm
+    secondary_variable = disp_x
```
it is a bit confusing we do not call `secondary_disp_x`

lindsayad replied:
I have made changes so that the param name is now `disp_x`

fdkong commented on modules/contact/test/tests/bouncing-block-contact/frictionless-weighted-gap-mixed-basis.i:
```diff
@@ -0,0 +1,173 @@
+  [./ncp_lm]
+    type = ApplyPenetrationConstraintLMMechanicalContact
+    secondary = 10
+    primary = 20
```
Should we make these names consistent?

```

    primary_boundary = 20
     secondary_boundary = 10
```

lindsayad replied:
So `ApplyPenetrationConstraintLMMechanicalContact` is actually a `NodeFaceConstraint` and it has has the `secondary` and `primary` params for many years. It is fine for them to do this because they don't also have lower-dimensional subdomains associated with their object like `MortarConstraints` do. So while the discrepancy is obnoxious I don't think we should change it at this point. I will note that our users of mortar contact (and contact in general) generally go through the `ContactAction` so this discrepancy is hidden from them

fdkong commented on modules/contact/test/tests/bouncing-block-contact/frictionless-weighted-gap-mixed-basis.i:
```diff
@@ -0,0 +1,173 @@
+    use_displaced_mesh = true
+    compute_lm_residuals = false
+  []
+  [normal_y]
```
I am trying to understand why we still need `normal_x` and `normal_y`.  I feel `weighted_gap_lm` and `ncp_lm` are enough

lindsayad replied:
The `ComputeWeightedGapLMMechanicalContact` and `ApplyPenetrationConstraintLMMechanicalContact` objects together determine the value of the LM, e.g. they compute the LM residual. However, they do not apply the LM (which represents the contact pressure) to the displacement residuals/equations. That is what `normal_x` and `normal_y` do. They apply the contact force/pressure represented by the LM to the displacement equations.

fdkong reviewed (changes requested):
It looks good to me. I am mainly confused by the constraint block of input files.

When applying the weighted gap formula, why we still need `NormalMortarMechanicalContact`?

lindsayad committed:
Fix potentially uninitialized warning in AutomaticMortarGeneration

lindsayad committed:
Use quiet NaN instead of numeric_limits::max

Use of `std::numeric_limits<Real>::max()` opened us up to the
possibilities of generating infinities any time we multiplied that
number by something greater than 1. This can easily lead to invalid
floating point operations which could result in unknown behavior when we
don't trap fpes. When we did trap FPEs we would get obviously error out
(this happened for some Hertzian contact problems that Antonio is
investigating with heavily curved surfaces in which plenty of secondary
nodes fail to project onto the primary surface, meaning we end up in the
code branch that formerly generated potential infs). Instead of opening
up the possibility of inf, I now insert a quiet NaN (such that we don't
create floating point exceptions for intermediate operations) and then
check for `std::isnan` when deciding whether to drive the LM or the gap
to zero in `ApplyPenetrationConstraintLMMechanicalContact`. The formerly
failing cases that Antonio gave me are now working with this patch

lindsayad commented on framework/src/constraints/AutomaticMortarGeneration.C:
```diff
@@ -774,7 +772,7 @@ AutomaticMortarGeneration::projectSecondaryNodesSinglePair(
             auto F = u(0) * nodal_normal(1) - u(1) * nodal_normal(0);
 
-            if (std::abs(F) < TOLERANCE)
+            if (std::abs(F) < TOLERANCE * TOLERANCE)
```
Haha it is a libMesh macro variable, but I will make this explicit

lindsayad replied:
I use non-squared `TOLERANCE` in other places

lindsayad committed:
Introduce member variables for tolerances in AutomaticMortarGeneration

In the future a developer can add APIs for changing these if they wish.
However, these tolerances have worked well for our mortar applications
thus far

lindsayad commented on modules/contact/doc/content/source/constraints/ApplyPenetrationConstraintLMMechanicalContact.md:
```diff
@@ -0,0 +1,33 @@
+\end{equation}
+
+where $g$ is a weighted gap and $\lambda$ is the contact pressure, a Lagrange multipler
+variable living on the secondary face. This object must always be used in
```
No, slave -> secondary, and master -> primary.

lindsayad commented on modules/contact/src/constraints/ApplyPenetrationConstraintLMMechanicalContact.C:
```diff
@@ -0,0 +1,205 @@
+  InputParameters params = NodeFaceConstraint::validParams();
+  params.set<bool>("use_displaced_mesh") = true;
+
+  params.addCoupledVar("disp_y", "The y displacement");
```
Good call. We actually do not need these parameters anymore

lindsayad commented on modules/contact/src/constraints/ApplyPenetrationConstraintLMMechanicalContact.C:
```diff
@@ -0,0 +1,205 @@
+      integrated_gap *= _c;
+
+      // If our LM is less than the integrated gap, then we return the LM, else we return 0 because
+      // the residual already is the weighted gap
```
> Did you mean the current residual vector already stored "the weighted gap"?

Correct!

lindsayad committed:
Address Fande review

lindsayad commented:
This'll teach me to add more sanity checks!!!

lindsayad committed:
Assert that we have nodes on our mortar mesh

lindsayad commented:
Ah, beautiful pass again. @fdkong this is ready for further review
